### PR TITLE
Add simple set of captured names to nested ParseNodeFnc

### PR DESCRIPTION
**Pull request by boingoing**
Each `ParseNodeFnc` has an IdentPtrSet which holds the set of names captured by the function. Captured names are ones which are referenced inside the function but not declared within the function.

Implement this by intercepting name references at PushPidRef and keep the name reference in the set of captured names for the current nested function. Remove any names from the captured name list which do not have dangling references after we've bound the pid ref stack for a nested function. When finished parsing a nested function, add the set of captured names onto the parent functions' captured name set.

Gate the feature via a ParseType flag and allow a forcing config option (`CreateParserState`).

Example captured name set:
```js
let x, y, z;
function foo() {
  function bar() {
    return x;
  }
  function baz(x) {
    return y;
  }
  function bat() {
    {
      let z;
      return z;
    }
  }
}

// captured name sets
// foo: [x, y]
// bar: [x]
// baz: [y]
// bat: []
```

### CODE REVIEW

**Comment by tcare** [comment]
I'm curious, how will we use this?

**Comment by pleath** [comment]
@tcare -- In a couple of ways. It allows us to skip scanning of previously scanned nested deferred functions and maintain correctness of stable closures. It also allows us to limit the work done on undeferral of those functions.

File: lib/Parser/Parse.cpp [hunk 1]
```diff
@@ -13815,8 +13827,84 @@ void ParseNode::Dump()
...
+            DumpCapturedNames(pnodeFnc, capturedNames, &this->m_nodeAllocator);
+            fflush(stdout);
+        }
+#endif
+    }
+}
+
+bool Parser::IsCreatingStateCache()
+{
+    return this->m_parseType == ParseType_StateCache
```
Comment (akroshg) [inline lib/Parser/Parse.cpp]: >this->m_parseType == ParseType_StateCache [](start = 11, length = 41)

curious what makes this true?

File: lib/Parser/Parse.cpp [hunk 2]
```diff
@@ -5650,6 +5652,8 @@ bool Parser::ParseFncDeclHelper(ParseNodeFnc * pnodeFnc, LPCOLESTR pNameHint, us
             CHAKRATEL_LANGSTATS_INC_LANGFEATURECOUNT(ES6, StrictModeFunction, m_scriptContext);
         }
 
+        ProcessCapturedNames(pnodeFnc);
```
Comment (akroshg) [inline lib/Parser/Parse.cpp]: we don;t do this for global function, right?

File: test/Basics/VerifyParserState.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+// Following lines generate a baseline with flags:
+// -ForceCreateParserState -trace:CreateParserState
+//
+// Uses WScript.LoadScript so that function numbering can be easier to reason about
+
+console.log('nested function capturing var decl');
```
Comment (akroshg) [inline test/Basics/VerifyParserState.js]: do we need to test with function with eval and with?

Comment (akroshg) [reply test/Basics/VerifyParserState.js]: and with function name expression? 

---
In reply to: [181448145](https://github.com/Microsoft/ChakraCore/pull/4983#discussion_r181448145) [](ancestors = 181448145)

Comment (tcare) [reply test/Basics/VerifyParserState.js]: Param scope would be good too since it takes different paths with pidrefstack

**Comment by tcare** [comment]
@pleath sounds great :) We could definitely improve some of the readability of the parser by using this too.

File: lib/Parser/Parse.cpp (hunk 1 above)
Comment (boingoing) [reply lib/Parser/Parse.cpp]: At the moment, nothing, which is why there is a force config flag. This flag would be pushed down from the API entrypoint which kicks-off the parse, like ParseScriptText

File: lib/Parser/Parse.cpp (hunk 2 above)
Comment (boingoing) [reply lib/Parser/Parse.cpp]: Yes, we don't do this for global functions.

File: test/Basics/VerifyParserState.js (hunk 3 above)
Comment (boingoing) [reply test/Basics/VerifyParserState.js]: Added unit tests for these cases, thanks for the feedback

File: lib/Parser/Parse.cpp [hunk 4]
```diff
@@ -13815,8 +13827,84 @@ void ParseNode::Dump()
...
+        IdentPtrSet* capturedNames = pnodeFnc->GetCapturedNames();
+        auto iter = capturedNames->GetIteratorWithRemovalSupport();
+
+        while (iter.IsValid())
+        {
+            const IdentPtr& pid = iter.CurrentValueReference();
+            PidRefStack* ref = pid->GetTopRef();
+
+            // If the pid has no refs left in our function's scope after binding, we didn't capture it.
+            if (!ref || ref->GetScopeId() < pnodeFnc->pnodeBodyScope->blockId)
```
[the remark this answers is not in the document]
Comment (boingoing) [reply lib/Parser/Parse.cpp]: I believe we should capture these as well. Made the change.

**Comment by boingoing** [comment]
@dotnet-bot test OSX static_osx_osx_debug please

**Comment by boingoing** [comment]
@dotnet-bot test OSX static_osx_osx_test please

**Pull request merged**